test/Test/Database/LSMTree/Internal/Serialise/Class.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Test.Database.LSMTree.Internal.Serialise.Class (tests) where
+
+import           Data.ByteString (ByteString)
+import           Data.ByteString.Lazy (LazyByteString)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Primitive (ByteArray)
+import           Data.Word
+import           Database.LSMTree.Extras.Generators ()
+import           Database.LSMTree.Internal.Serialise
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Serialise.Class"
+    [ allProperties @Word64 "Word64"
+    , allProperties @ByteString "ByteString"
+    , allProperties @LazyByteString "LazyByteString"
+    , allProperties @ShortByteString "ShortByteString"
+    , allProperties @ByteArray "ByteArray"
+    ]
+
+allProperties :: forall a. (Ord a, Show a, Arbitrary a, SerialiseKey a, SerialiseValue a) => String -> TestTree
+allProperties name =
+    testGroup name
+      [ testProperty "prop_roundtripSerialiseKey" $
+          prop_roundtripSerialiseKey @a
+      , testProperty "prop_roundtripSerialiseValue" $
+          prop_roundtripSerialiseValue @a
+      , testProperty "prop_orderPreservationSerialiseKey" $
+          prop_orderPreservationSerialiseKey @a
+      , testProperty "prop_orderPreservationSerialiseValue" $
+          prop_orderPreservationSerialiseValue @a
+      ]
+
+prop_roundtripSerialiseKey :: (Eq k, Show k, SerialiseKey k) => k -> Property
+prop_roundtripSerialiseKey k =
+    deserialiseKey (serialiseKey k) === k
```

Comment:
We have serialiseKeyIdentity serialiseKeyPreservesOrdering, sadly there are somewhat useless in tests, aren't they?

Reply:
Ah, how did I miss them? I could use `serialiseValueIdentity`, it just doesn't print the mismatch as `===` does. And I should probably add a test for `serialiseValueConcatDistributes` (or something like it).

Reply:
They are *yet* unused. My intent here was to expose these functions as part of the public API, so users can check whether their serialisation functions are good. This is similar to https://hackage.haskell.org/package/serialise-0.2.6.1/docs/Codec-Serialise-Properties.html

Reply:
> Ah, how did I miss them? I could use `serialiseValueIdentity`, it just doesn't print the mismatch as `===` does. And I should probably add a test for `serialiseValueConcatDistributes` (or something like it).

If you want to redefine the properties in terms of `===` that should be fine. Maybe add an assertion that checks wether `serialiseKeyIdentity` and this property agree?

Reply:
I used the predefined properties and just added a few calls to `counterexample`.

The only property where I didn't use a simple wrapper is `serialiseValueConcatDistributes`. I don't test it on arbitrary `RawBytes`, but just on valid serialisation output.
One could argue I should use `[RawBytes]` as input and either require deserialisation to be total or catch `error` calls, but it's simpler and IMO okay to use valid input only.